modules/gapi/src/executor/gstreamingexecutor.cpp

```diff
@@ -1380,11 +1382,24 @@ cv::gimpl::GStreamingExecutor::GStreamingExecutor(std::unique_ptr<ade::Graph> &&
     });
 }
 
-cv::gimpl::GStreamingExecutor::~GStreamingExecutor()
-{
+void cv::gimpl::GStreamingExecutor::destruct() {
     if (state == State::READY || state == State::RUNNING)
         stop();
 }
+cv::gimpl::GStreamingExecutor::~GStreamingExecutor() noexcept(false)
+{
+    if (std::uncaught_exception()) {
+        try {
+            destruct();
+        } catch (const std::logic_error& e) {
+            std::stringstream message;
+            message << "~GStreamingExecutor() threw exception with message '" << e.what() << "'\n";
+            GAPI_LOG_WARNING(NULL, message.str());
+        }
+    } else {
+        destruct();
```

Comment:
destruct throws error?

I checked code and found root cause of exception
```
    // Pull messages from the final queue to ensure completion
    Cmd cmd;
    while (!cv::util::holds_alternative<Stop>(cmd))
    {
        m_out_queue.pop(cmd);
    }
    GAPI_Assert(cv::util::holds_alternative<Stop>(cmd));   <----
```

And 
1) Assert looks redundant, because while-loop won't breaks otherwise
2) this may be solved by more simplest way then using std::uncaught_exception
a) separate stop() on stop() & wait_stop/shutdown(), and just call stop() in ~dtor because it looks like you should not extracts some cmd (because currently we're dropping them anyway)
b) do not throw Assert exception at all

From my perspective `std::uncaught_exception` is redundant and stirred ups and should be used in more dead-ended situations. But `stop` should not throw anything

Reply:
I think your comment is highly reasonable, firstly about freeing destruction from exceptions: I'd rather dive a bit deeper & solve as many rootcauses as possible than simply cover all with try-catch.

Secondly, I like this `stop()` separation suggestion.

Reply:
Though, there are much more places where destructor can fall over.
The most obvious is assert in `push()` of `concurrent_bounded_queue`:
```
   if (m_capacity && m_capacity == m_data.size()) {
        // if there is a limit and it is reached, wait
        m_cond_full.wait(lock, [&](){return m_capacity > m_data.size();});
        GAPI_Assert(m_capacity > m_data.size());
    }
```
and maybe some others.

@dmatveev what do you think on this destructor issue?

Reply:
@OrestChura  i see you point, but: from my perspective `bounded_queue` should not throws too because `std::conditional_variable` here retain its postcondition after `.wait`  - so we do not need to debug standard c++ library and this check looks impossible according to `bounded_queue` design. So this throwing exception reason MUST be predictable, because this code doesn't interact with some user input or depends some OS/IO/RAM primitives setting but all behavior ensured by STD. I mean that it cannot throw sporadically and 100% determined from current code logic from 100% determined cases. 
SO, there are should have been unit test for that "case" which ensure that exception must or must not to ensure guarantee that `bounded_queue` is consistent at all. But I'm pretty sure that it is useless here

What's about other exception: `std::bad_alloc` and so on - they are uncontrollable/unpredictable from us usually and/or based on user input or other system settings and SHOULD be checked in appropriate places OR ignored and falls into destructor and let program terminate in usual way, If it doesn't contradict with current policy

Example of predictable & controllable behavior:

```
try {
cin >> size; // user input `-1`
char* new char [size];  // usually throws std::bad_alloc on `-1` 
} catch (...)
{
cerr << "Too much size requested: " << size << std::endl;
}
```

Example of non predictable behavior
```
std::string aaa = "aaa";
foo(aaa);

foo (std::string bbb)  <---- copy may throw
```

The second example usually in not handled in most common applications and should be ignored

In result i assume that `coverity` doesn't dive into STD code and doesn't check throwing from vector push and so on and it's limited by application code only


 
**P.S.** I found a bug in bounded queue here. because `notify_one` must be `notify_all` and the next part should be changed

```
    while (m_capacity && m_capacity == m_data.size()) {
            // if there is a limit and it is reached, wait
           m_cond_full.wait(lock, [&](){return m_capacity > m_data.size();});
    }
```

Let's create an issue